vehicle-simulator/main.go

```diff
@@ -0,0 +1,186 @@
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log/slog"
+	"os"
+	"os/signal"
+	"syscall"
+
+	dittoModel "github.com/eclipse/ditto-clients-golang/model"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/eclipse/ditto-clients-golang/protocol/things"
+	mqtt "github.com/eclipse/paho.mqtt.golang"
+)
+
+type DeviceData struct {
+	DeviceID string `json:"deviceID"`
+	TenantID string `json:"tenantID"`
+	PolicyID string `json:"policyID"`
+}
+
+var mqttClient mqtt.Client
+var deviceData = DeviceData{}
+var telemetrySenderStarted bool
+var selectedConnectorDeviceID string // holds the last deviceID from the previous active connector
+var ctdSubscriptionTopics []string
+
+var stopTelemetrySender = make(chan struct{})
+var controlTelemetrySender = make(chan string)
+
+// Define flag variables
+var ctdEffectDelay = flag.Int("ctd-effect-delay", 30, "In seconds, Configure trunk/alarms state time")
+var periodicUpdatesDelay = flag.Int("periodic-updates-delay", 5, "In seconds, Configure periodic updates (minimum 5)")
+var idleStateTime = flag.Int("idle-state-time", 120, "In seconds, Configure when the system will enter idle state. Idle state will send telemetry data infrequently. Receiving CTD command will reset the idle state.")
+var initialPeriodicUpdatesDelay = 0 // set the initial periodic updates delay in case of idle state active
+
+func main() {
+	flag.Parse()
+	validateFlags()
+
+	keepAlive := make(chan os.Signal, 1)
+	signal.Notify(keepAlive, os.Interrupt, syscall.SIGTERM)
+
+	initializeClient()
+	getDeviceData()
+
+	defer func() { stopTelemetrySender <- struct{}{} }()
+	defer mqttClient.Disconnect(1000)
+
+	<-keepAlive
+}
+
+func initializeClient() {
+	brokerURL := fmt.Sprintf("tcp://%s:%s", getEnvVariable("ADAPTER_BROKER_ADDRESS", "localhost"), getEnvVariable("ADAPTER_BROKER_PORT", "1883"))
+	slog.Info("Broker configuration", "URL", brokerURL)
+	options := mqtt.NewClientOptions()
+	options.AddBroker(brokerURL)
+	mqttClient = mqtt.NewClient(options)
+
+	if token := mqttClient.Connect(); token.Wait() && token.Error() != nil {
+		panic(token.Error())
+	}
+}
+
+func getEnvVariable(variableName string, defaultValue string) string {
+	value := os.Getenv(variableName)
+	if len(value) == 0 {
+		return defaultValue
+	}
+	return value
+}
+
+func getDeviceData() {
+	mqttClient.Subscribe("edge/thing/response", 1, handleDeviceDataReceived).Wait()
+	mqttClient.Publish("edge/thing/request", 1, false, "{}").Wait()
+}
+
+func handleDeviceDataReceived(sourceClient mqtt.Client, message mqtt.Message) {
+	if err := json.Unmarshal(message.Payload(), &deviceData); err != nil {
```

Comment:
Use temporary DeviceData structure. Keep _deviceData_ variable for the current device.